Add parsing of `classify` function in mamlscript

alkamin opened the pull request:
## Overview

This PR adds support for a `classify` function in *mamlscript*.

Syntactically, this is really a first pass but it is pretty straightforward:

`classify(input_expression, classifications)`

- `input_expression` is any valid *mamlscript* expression. Ex: `A + B + C`
- `classifications` is an object mapping the maximum value of a class (double as string) to an output value (integer). Ex: `{ "0.0": 0, "1.0" : 1}`

Additionally, this PR also adds like-operation merging. Previously, an expression like `A + B + C` would have been parsed into a tree more like `A + (B + C)`. This would stack, so expressions could end up unnecessarily deep (`A + (B + (C + (D + E)))`). This is accomplished in a way that respects parenthetical groupings.

### Checklist

~- [ ] Styleguide updated, if necessary~
~- [ ] Swagger specification updated, if necessary~
~- [ ] Symlinks from new migrations present or corrected for any new migrations~
- [x] PR has a name that won't get you publicly shamed for vagueness

### Demo

`classify((NIR - R)/(NIR + R), {"0.0": 0, "1.0": 1})`

![image](https://user-images.githubusercontent.com/2442245/29927023-1ad49fee-8e33-11e7-8b97-0ce4d4e0cc75.png)


`A + B + C + (D + E)`

![image](https://user-images.githubusercontent.com/2442245/29926971-eb5d09b8-8e32-11e7-9dcb-58cf64597ac1.png)

### Notes

We'll soon be moving this work to a separate library so that it can be made more robust and tested. (https://github.com/raster-foundry/maml.js)

## Testing Instructions

* Make an expression with the classify function and see that it works
* Make an expression with repeated operations (A + B + C) and see that they are combined

Closes #2448 
Closes #2367

ddohler commented on app-frontend/src/app/components/tools/toolCreateModal/toolCreateModal.controller.js:
```diff
@@ -46,7 +46,7 @@ export default class ToolCreateModalController {
         let expressionTree = null;
         try {
             expressionTree = mathjs.parse(this.definitionExpression);
-            this.toolBuffer.definition = this.transformNode(expressionTree);
```
Looks like this function still exists but is no longer used -- should it be removed?

ddohler commented on app-frontend/src/app/components/tools/toolCreateModal/toolCreateModal.controller.js:
```diff
@@ -126,6 +126,84 @@ export default class ToolCreateModalController {
         return builtNode;
     }
 
+    expressionTreeToMAML(tree, value, allowCollapse = true) {
```
This could use some commenting to provide guideposts to a later reader; the code looks fine as-is, but without a deep knowledge of both MAML and mathjs it's tough to understand why it does some of the things that it does.

alkamin commented:
I added brace balancing and removed the unused function.

Avoiding the quotes would involve expression pre-processing which is in the plan for maml.js, so I'm going to hold off for now.
